src/librustc/mir/interpret/allocation.rs

```diff
@@ -101,18 +117,62 @@ impl<Tag> Allocation<Tag> {
     }
 
     pub fn undef(size: Size, align: Align) -> Self {
+        // Ensure initialization later gives the correct length.
         assert_eq!(size.bytes() as usize as u64, size.bytes());
         Allocation {
-            bytes: vec![0; size.bytes() as usize],
+            alloc: None,
             relocations: Relocations::new(),
-            undef_mask: UndefMask::new(size, false),
+            size,
             align,
             mutability: Mutability::Mutable,
             extra: (),
         }
     }
 }
 
+/// Raw accessors. Provide compatibility through lazy allocation of undef.
+impl<Tag, Extra> Allocation<Tag, Extra> {
+    pub fn len(&self) -> usize {
+        self.alloc.as_ref()
+            .map(|alloc| {
+                debug_assert_eq!(alloc.bytes.len() as u64, self.size.bytes());
+                alloc.bytes.len()
+            })
+            .unwrap_or(self.size.bytes() as usize)
+    }
+
+    /// Access without going through the logical part of the allocation.
+    ///
+    /// Interpreter computation should not depend on the content of these bytes and caller has the
+    /// ensure the range is valid.
+    pub fn get_raw_bytes(&self, range: Range<usize>) -> BytesOrElidedUndef<'_> {
```

Comment:
What is "the logical part of the allocation"? Inf act I don't understand any part of this doc comment.

Reply:
I wanted to say that this ignores the `UndefMask`, relocations, and range checks. It probably should have said exactly that:

```
    /// Directly access the underlying bytes of an allocation.
    ///
    /// This ignores the undef mask, relocations, and requires the caller to have made proper
    /// bounds checks. Interpreter computation should not depend on the content of these bytes.
```